### PR TITLE
fix: make deleteOldFiles / moveNewVersionInPlace more robust

### LINKED ISSUE #158: Updater fails to complete (step Delete old files) / `Could not rmdir`
I believe issue https://github.com/nextcloud/server/issues/2737 is not properly solved yet in NC13.0.0.
See discussions in above issue.

Summary:
 - The step "Delete old files" fails with a message like "Could not rmdir: /home/***/www/***/nc-test/nextcloud/updater/../core/doc/admin/_images".
- Pressing retry helps but must be done a dozen times (update will finally complete).
- This issue is known to occur reproducibly on a FreeBSD 10.3 on a virtual server of the hoster hostpoint.ch, hos ever it might be that it also occurs on other BSDs (See comment of a user from Canada: https://github.com/nextcloud/server/issues/2737#issuecomment-362788491) .

To trace it further down, I added the lines 
```
                $this->log("## recursiveDelete: deleting folder $folder");
		foreach ($directories as $dir) {
			$this->log("#### recursiveDelete: $dir");
		}
```
after
```
		foreach ($files as $file) {
			unlink($file);
		}
``` 
in the function `recursiveDelete` in updater/index.php.
This gives the following updater.log (shortened) below.
The `recursiveDelete` seems to work fine for many many times, but then suddenly fails with the message
```
2018-02-07T22:11:38+0100 Nv5wo8Bsbf [error] POST request failed with other exception
2018-02-07T22:11:38+0100 Nv5wo8Bsbf [error] Exception: Exception
Message: Could not rmdir: /home/***/www/***/nc-test/nextcloud/updater/../core/doc/admin/_images
Code:0
Trace:
#0 /home/***/www/***/nc-test/nextcloud/updater/index.php(1316): Updater->deleteOldFiles()
#1 {main}
File:/home/***/www/***/nc-test/nextcloud/updater/index.php
Line:943
```
Against my initial thought, this seems not to be caused by a `rmdir` failure!

<details><summary>updater.log</summary>

```

2018-02-07T22:03:49+0100 NPXnWlJMbo [info] request to updater
2018-02-07T22:03:49+0100 NPXnWlJMbo [info] currentStep()
2018-02-07T22:03:49+0100 NPXnWlJMbo [info] show HTML page
2018-02-07T22:03:49+0100 NPXnWlJMbo [info] current version: 13.0.0 RC 1 build time: 2018-01-11T19:40:31+00:00 4f30c2fbe43e0d6f24ba3bff881526b2b6c2747d
2018-02-07T22:03:49+0100 NPXnWlJMbo [info] getUpdateServerResponse()
2018-02-07T22:03:49+0100 NPXnWlJMbo [info] updaterServer: https://updates.nextcloud.org/updater_server/
2018-02-07T22:03:49+0100 NPXnWlJMbo [info] releaseChannel: beta
2018-02-07T22:03:49+0100 NPXnWlJMbo [info] internal version: 13.0.0.10
2018-02-07T22:03:49+0100 NPXnWlJMbo [info] updateURL: https://updates.nextcloud.org/updater_server/?version=13x0x0x10xxxbetaxx2018-01-11T19%3A40%3A31%2B00%3A00+4f30c2fbe43e0d6f24ba3bff881526b2b6c2747dx5x6x33
2018-02-07T22:03:49+0100 NPXnWlJMbo [info] getUpdateServerResponse response: Array
(
    [version] => 13.0.0.14
    [versionstring] => Nextcloud 13.0.0
    [url] => https://download.nextcloud.com/server/releases/nextcloud-13.0.0.zip
    [web] => https://docs.nextcloud.com/server/12/admin_manual/maintenance/upgrade.html
    [autoupdater] => 1
    [signature] => q5bqFR15JaWW6oGM+iNtC7JoDsRHPXAkySJ5TMgcbAfajheMUwgQtytK2S1vpJd7
UbhO5CfWStBQKFNR6/tV6R++xLbIBZgXhvjLtpciug+dNBMANZNUiqEbdYLZSWEp
Y3HOk087s3o0wazQ/kLDmFmCzW8bngpcI1rRDJTXCS6uf/0BVatOaIoByJRkArnw
ir+Hd8swyREt3jrngeePu6/ZrB+5toGcEHaSCmTNwJ7ipKnwi3mPP0XcGWJVswzY
WpeJUBR9OkzLQ8Y6sKIEOZrDBsoSjFp0YN6Adgbxgbd4UPwgaJVOFdfuW18RNWWw
Tx/vStIU+zA9/Yan/RotKg==
)

2018-02-07T22:03:49+0100 NPXnWlJMbo [info] checkForUpdate() Array
(
    [version] => 13.0.0.14
    [versionstring] => Nextcloud 13.0.0
    [url] => https://download.nextcloud.com/server/releases/nextcloud-13.0.0.zip
    [web] => https://docs.nextcloud.com/server/12/admin_manual/maintenance/upgrade.html
    [autoupdater] => 1
    [signature] => q5bqFR15JaWW6oGM+iNtC7JoDsRHPXAkySJ5TMgcbAfajheMUwgQtytK2S1vpJd7
UbhO5CfWStBQKFNR6/tV6R++xLbIBZgXhvjLtpciug+dNBMANZNUiqEbdYLZSWEp
Y3HOk087s3o0wazQ/kLDmFmCzW8bngpcI1rRDJTXCS6uf/0BVatOaIoByJRkArnw
ir+Hd8swyREt3jrngeePu6/ZrB+5toGcEHaSCmTNwJ7ipKnwi3mPP0XcGWJVswzY
WpeJUBR9OkzLQ8Y6sKIEOZrDBsoSjFp0YN6Adgbxgbd4UPwgaJVOFdfuW18RNWWw
Tx/vStIU+zA9/Yan/RotKg==
)

2018-02-07T22:03:49+0100 NPXnWlJMbo [info] end of checkForUpdate() Update to Nextcloud 13.0.0 available. (channel: "beta")<br /><span class="light">Following file will be downloaded automatically:</span> <code class="light">https://download.nextcloud.com/server/releases/nextcloud-13.0.0.zip</code>
2018-02-07T22:05:51+0100 fE2PeuhpO0 [info] request to updater
2018-02-07T22:05:51+0100 fE2PeuhpO0 [info] currentStep()
2018-02-07T22:05:51+0100 fE2PeuhpO0 [info] POST request for step "1"
2018-02-07T22:05:51+0100 fE2PeuhpO0 [info] startStep("1")
2018-02-07T22:05:51+0100 fE2PeuhpO0 [info] checkForExpectedFilesAndFolders()
2018-02-07T22:05:51+0100 fE2PeuhpO0 [info] end of checkForExpectedFilesAndFolders()
2018-02-07T22:05:51+0100 fE2PeuhpO0 [info] endStep("1")
2018-02-07T22:05:51+0100 JQw86bhs34 [info] request to updater
2018-02-07T22:05:51+0100 JQw86bhs34 [info] currentStep()
2018-02-07T22:05:51+0100 JQw86bhs34 [info] Step 1 is in state "end".
2018-02-07T22:05:51+0100 JQw86bhs34 [info] POST request for step "2"
2018-02-07T22:05:51+0100 JQw86bhs34 [info] startStep("2")
2018-02-07T22:05:51+0100 JQw86bhs34 [info] checkWritePermissions()
2018-02-07T22:05:51+0100 JQw86bhs34 [info] end of checkWritePermissions()
2018-02-07T22:05:51+0100 JQw86bhs34 [info] endStep("2")
2018-02-07T22:05:51+0100 h1rdmb0yzV [info] request to updater
2018-02-07T22:05:51+0100 h1rdmb0yzV [info] currentStep()
2018-02-07T22:05:51+0100 h1rdmb0yzV [info] Step 2 is in state "end".
2018-02-07T22:05:51+0100 h1rdmb0yzV [info] POST request for step "3"
2018-02-07T22:05:51+0100 h1rdmb0yzV [info] startStep("3")
2018-02-07T22:05:51+0100 h1rdmb0yzV [info] setMaintenanceMode("true")
2018-02-07T22:05:51+0100 h1rdmb0yzV [info] configFileName /home/***/www/***/nc-test/nextcloud/updater/../config/config.php
2018-02-07T22:05:52+0100 h1rdmb0yzV [info] end of setMaintenanceMode()
2018-02-07T22:05:52+0100 h1rdmb0yzV [info] endStep("3")
2018-02-07T22:05:52+0100 MPAPEPjVe7 [info] request to updater
2018-02-07T22:05:52+0100 MPAPEPjVe7 [info] currentStep()
2018-02-07T22:05:52+0100 MPAPEPjVe7 [info] Step 3 is in state "end".
2018-02-07T22:05:52+0100 MPAPEPjVe7 [info] POST request for step "4"
2018-02-07T22:05:52+0100 MPAPEPjVe7 [info] startStep("4")
2018-02-07T22:05:52+0100 MPAPEPjVe7 [info] createBackup()
2018-02-07T22:09:31+0100 MPAPEPjVe7 [info] end of createBackup()
2018-02-07T22:09:31+0100 MPAPEPjVe7 [info] endStep("4")
2018-02-07T22:09:31+0100 WXYs64DnxH [info] request to updater
2018-02-07T22:09:31+0100 WXYs64DnxH [info] currentStep()
2018-02-07T22:09:31+0100 WXYs64DnxH [info] Step 4 is in state "end".
2018-02-07T22:09:31+0100 WXYs64DnxH [info] POST request for step "5"
2018-02-07T22:09:31+0100 WXYs64DnxH [info] startStep("5")
2018-02-07T22:09:32+0100 WXYs64DnxH [info] downloadUpdate()
2018-02-07T22:09:32+0100 WXYs64DnxH [info] getUpdateServerResponse()
2018-02-07T22:09:32+0100 WXYs64DnxH [info] updaterServer: https://updates.nextcloud.org/updater_server/
2018-02-07T22:09:32+0100 WXYs64DnxH [info] releaseChannel: beta
2018-02-07T22:09:32+0100 WXYs64DnxH [info] internal version: 13.0.0.10
2018-02-07T22:09:32+0100 WXYs64DnxH [info] updateURL: https://updates.nextcloud.org/updater_server/?version=13x0x0x10xxxbetaxx2018-01-11T19%3A40%3A31%2B00%3A00+4f30c2fbe43e0d6f24ba3bff881526b2b6c2747dx5x6x33
2018-02-07T22:09:32+0100 WXYs64DnxH [info] getUpdateServerResponse response: Array
(
    [version] => 13.0.0.14
    [versionstring] => Nextcloud 13.0.0
    [url] => https://download.nextcloud.com/server/releases/nextcloud-13.0.0.zip
    [web] => https://docs.nextcloud.com/server/12/admin_manual/maintenance/upgrade.html
    [autoupdater] => 1
    [signature] => q5bqFR15JaWW6oGM+iNtC7JoDsRHPXAkySJ5TMgcbAfajheMUwgQtytK2S1vpJd7
UbhO5CfWStBQKFNR6/tV6R++xLbIBZgXhvjLtpciug+dNBMANZNUiqEbdYLZSWEp
Y3HOk087s3o0wazQ/kLDmFmCzW8bngpcI1rRDJTXCS6uf/0BVatOaIoByJRkArnw
ir+Hd8swyREt3jrngeePu6/ZrB+5toGcEHaSCmTNwJ7ipKnwi3mPP0XcGWJVswzY
WpeJUBR9OkzLQ8Y6sKIEOZrDBsoSjFp0YN6Adgbxgbd4UPwgaJVOFdfuW18RNWWw
Tx/vStIU+zA9/Yan/RotKg==
)

2018-02-07T22:09:35+0100 WXYs64DnxH [info] end of downloadUpdate()
2018-02-07T22:09:35+0100 WXYs64DnxH [info] endStep("5")
2018-02-07T22:09:35+0100 rOIT2543DE [info] request to updater
2018-02-07T22:09:35+0100 rOIT2543DE [info] currentStep()
2018-02-07T22:09:35+0100 rOIT2543DE [info] Step 5 is in state "end".
2018-02-07T22:09:35+0100 rOIT2543DE [info] POST request for step "6"
2018-02-07T22:09:35+0100 rOIT2543DE [info] startStep("6")
2018-02-07T22:09:35+0100 rOIT2543DE [info] verifyIntegrity()
2018-02-07T22:09:35+0100 rOIT2543DE [info] getUpdateServerResponse()
2018-02-07T22:09:35+0100 rOIT2543DE [info] updaterServer: https://updates.nextcloud.org/updater_server/
2018-02-07T22:09:35+0100 rOIT2543DE [info] releaseChannel: beta
2018-02-07T22:09:35+0100 rOIT2543DE [info] internal version: 13.0.0.10
2018-02-07T22:09:35+0100 rOIT2543DE [info] updateURL: https://updates.nextcloud.org/updater_server/?version=13x0x0x10xxxbetaxx2018-01-11T19%3A40%3A31%2B00%3A00+4f30c2fbe43e0d6f24ba3bff881526b2b6c2747dx5x6x33
2018-02-07T22:09:35+0100 rOIT2543DE [info] getUpdateServerResponse response: Array
(
    [version] => 13.0.0.14
    [versionstring] => Nextcloud 13.0.0
    [url] => https://download.nextcloud.com/server/releases/nextcloud-13.0.0.zip
    [web] => https://docs.nextcloud.com/server/12/admin_manual/maintenance/upgrade.html
    [autoupdater] => 1
    [signature] => q5bqFR15JaWW6oGM+iNtC7JoDsRHPXAkySJ5TMgcbAfajheMUwgQtytK2S1vpJd7
UbhO5CfWStBQKFNR6/tV6R++xLbIBZgXhvjLtpciug+dNBMANZNUiqEbdYLZSWEp
Y3HOk087s3o0wazQ/kLDmFmCzW8bngpcI1rRDJTXCS6uf/0BVatOaIoByJRkArnw
ir+Hd8swyREt3jrngeePu6/ZrB+5toGcEHaSCmTNwJ7ipKnwi3mPP0XcGWJVswzY
WpeJUBR9OkzLQ8Y6sKIEOZrDBsoSjFp0YN6Adgbxgbd4UPwgaJVOFdfuW18RNWWw
Tx/vStIU+zA9/Yan/RotKg==
)

2018-02-07T22:09:35+0100 rOIT2543DE [info] storage location: /home/***/www/***/nc-test/data/updater-occj10ln2to7/downloads/
2018-02-07T22:09:38+0100 rOIT2543DE [info] end of verifyIntegrity()
2018-02-07T22:09:38+0100 rOIT2543DE [info] endStep("6")
2018-02-07T22:09:38+0100 Yqtzg7oz3C [info] request to updater
2018-02-07T22:09:38+0100 Yqtzg7oz3C [info] currentStep()
2018-02-07T22:09:38+0100 Yqtzg7oz3C [info] Step 6 is in state "end".
2018-02-07T22:09:38+0100 Yqtzg7oz3C [info] POST request for step "7"
2018-02-07T22:09:38+0100 Yqtzg7oz3C [info] startStep("7")
2018-02-07T22:09:38+0100 Yqtzg7oz3C [info] extractDownload()
2018-02-07T22:09:38+0100 Yqtzg7oz3C [info] storage location: /home/***/www/***/nc-test/data/updater-occj10ln2to7/downloads/
2018-02-07T22:11:03+0100 Yqtzg7oz3C [info] end of extractDownload()
2018-02-07T22:11:03+0100 Yqtzg7oz3C [info] endStep("7")
2018-02-07T22:11:03+0100 HZAFrGK53h [info] request to updater
2018-02-07T22:11:03+0100 HZAFrGK53h [info] currentStep()
2018-02-07T22:11:03+0100 HZAFrGK53h [info] Step 7 is in state "end".
2018-02-07T22:11:03+0100 HZAFrGK53h [info] POST request for step "8"
2018-02-07T22:11:03+0100 HZAFrGK53h [info] startStep("8")
2018-02-07T22:11:03+0100 HZAFrGK53h [info] replaceEntryPoints()
2018-02-07T22:11:03+0100 HZAFrGK53h [info] replace index.php
2018-02-07T22:11:04+0100 HZAFrGK53h [info] replace status.php
2018-02-07T22:11:04+0100 HZAFrGK53h [info] replace remote.php
2018-02-07T22:11:04+0100 HZAFrGK53h [info] replace public.php
2018-02-07T22:11:04+0100 HZAFrGK53h [info] replace ocs/v1.php
2018-02-07T22:11:04+0100 HZAFrGK53h [info] replace ocs/v2.php
2018-02-07T22:11:04+0100 HZAFrGK53h [info] end of replaceEntryPoints()
2018-02-07T22:11:04+0100 HZAFrGK53h [info] endStep("8")
2018-02-07T22:11:04+0100 Nv5wo8Bsbf [info] request to updater
2018-02-07T22:11:04+0100 Nv5wo8Bsbf [info] currentStep()
2018-02-07T22:11:04+0100 Nv5wo8Bsbf [info] Step 8 is in state "end".
2018-02-07T22:11:04+0100 Nv5wo8Bsbf [info] POST request for step "9"
2018-02-07T22:11:04+0100 Nv5wo8Bsbf [info] startStep("9")
2018-02-07T22:11:04+0100 Nv5wo8Bsbf [info] deleteOldFiles()
2018-02-07T22:11:05+0100 Nv5wo8Bsbf ## recursiveDelete: deleting folder /home/***/www/***/nc-test/nextcloud/updater/../apps/activity
2018-02-07T22:11:05+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/activity/js
2018-02-07T22:11:05+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/activity/l10n/.tx
2018-02-07T22:11:05+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/activity/l10n
2018-02-07T22:11:05+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/activity/templates/settings
2018-02-07T22:11:05+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/activity/templates
2018-02-07T22:11:05+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/activity/css
2018-02-07T22:11:05+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/activity/img
2018-02-07T22:11:05+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/activity/lib/Formatter
2018-02-07T22:11:05+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/activity/lib/Settings
2018-02-07T22:11:05+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/activity/lib/Command
2018-02-07T22:11:05+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/activity/lib/Parameter
2018-02-07T22:11:05+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/activity/lib/AppInfo
2018-02-07T22:11:05+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/activity/lib/Migration
2018-02-07T22:11:05+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/activity/lib/Exception
2018-02-07T22:11:05+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/activity/lib/Filter
2018-02-07T22:11:05+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/activity/lib/Extension
2018-02-07T22:11:05+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/activity/lib/Controller
2018-02-07T22:11:05+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/activity/lib/BackgroundJob
2018-02-07T22:11:05+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/activity/lib
2018-02-07T22:11:05+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/activity/appinfo
2018-02-07T22:11:05+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/activity/docs
[..]
2018-02-07T22:11:34+0100 Nv5wo8Bsbf ## recursiveDelete: deleting folder /home/***/www/***/nc-test/nextcloud/updater/../apps/user_ldap
2018-02-07T22:11:34+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/user_ldap/js/wizard
2018-02-07T22:11:34+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/user_ldap/js
2018-02-07T22:11:34+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/user_ldap/vendor/ui-multiselect/src
2018-02-07T22:11:34+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/user_ldap/vendor/ui-multiselect
2018-02-07T22:11:34+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/user_ldap/vendor
2018-02-07T22:11:34+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/user_ldap/l10n
2018-02-07T22:11:34+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/user_ldap/templates
2018-02-07T22:11:34+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/user_ldap/css
2018-02-07T22:11:34+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/user_ldap/img
2018-02-07T22:11:34+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/user_ldap/ajax
2018-02-07T22:11:34+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/user_ldap/composer/composer
2018-02-07T22:11:34+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/user_ldap/composer
2018-02-07T22:11:34+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/user_ldap/lib/Settings
2018-02-07T22:11:34+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/user_ldap/lib/Command
2018-02-07T22:11:34+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/user_ldap/lib/AppInfo
2018-02-07T22:11:34+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/user_ldap/lib/Migration
2018-02-07T22:11:34+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/user_ldap/lib/User
2018-02-07T22:11:34+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/user_ldap/lib/Jobs
2018-02-07T22:11:34+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/user_ldap/lib/Notification
2018-02-07T22:11:34+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/user_ldap/lib/Mapping
2018-02-07T22:11:34+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/user_ldap/lib/Exceptions
2018-02-07T22:11:34+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/user_ldap/lib/Controller
2018-02-07T22:11:34+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/user_ldap/lib
2018-02-07T22:11:34+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/user_ldap/appinfo
2018-02-07T22:11:34+0100 Nv5wo8Bsbf ## recursiveDelete: deleting folder /home/***/www/***/nc-test/nextcloud/updater/../apps/workflowengine
2018-02-07T22:11:34+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/workflowengine/js
2018-02-07T22:11:34+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/workflowengine/l10n
2018-02-07T22:11:34+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/workflowengine/templates
2018-02-07T22:11:34+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/workflowengine/css
2018-02-07T22:11:34+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/workflowengine/lib/Settings
2018-02-07T22:11:34+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/workflowengine/lib/AppInfo
2018-02-07T22:11:34+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/workflowengine/lib/Check
2018-02-07T22:11:34+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/workflowengine/lib/Controller
2018-02-07T22:11:34+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/workflowengine/lib
2018-02-07T22:11:34+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/apps/workflowengine/appinfo
2018-02-07T22:11:35+0100 Nv5wo8Bsbf [info] config sample exists
2018-02-07T22:11:35+0100 Nv5wo8Bsbf [info] themes README exists
2018-02-07T22:11:35+0100 Nv5wo8Bsbf ## recursiveDelete: deleting folder /home/***/www/***/nc-test/nextcloud/updater/../themes/example/
2018-02-07T22:11:35+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/themes/example/core/css
2018-02-07T22:11:35+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/themes/example/core/img
2018-02-07T22:11:35+0100 Nv5wo8Bsbf #### recursiveDelete: /home/***/www/***/nc-test/nextcloud/themes/example/core
2018-02-07T22:11:38+0100 Nv5wo8Bsbf [error] POST request failed with other exception
2018-02-07T22:11:38+0100 Nv5wo8Bsbf [error] Exception: Exception
Message: Could not rmdir: /home/***/www/***/nc-test/nextcloud/updater/../core/doc/admin/_images
Code:0
Trace:
#0 /home/***/www/***/nc-test/nextcloud/updater/index.php(1316): Updater->deleteOldFiles()
#1 {main}
File:/home/***/www/***/nc-test/nextcloud/updater/index.php
Line:943

2018-02-07T22:11:38+0100 Nv5wo8Bsbf [info] rollbackChanges("9")
2018-02-07T22:11:38+0100 Nv5wo8Bsbf [info] unlink .step
2018-02-07T22:11:38+0100 Nv5wo8Bsbf [info] rollbackChanges - step >= 7
2018-02-07T22:11:38+0100 Nv5wo8Bsbf [info] end of  rollbackChanges()
```

</details>

## Server configuration detail

**Operating system:** FreeBSD 10.3-RELEASE-p24 FreeBSD 10.3-RELEASE-p24 #14 r326012M: Mon Nov 20 12:42:55 UTC 2017     root@port10.adm.hostpoint.ch:/usr/obj/usr/src/sys/HOSTPOINT amd64

**Webserver:** Apache/2.4 (cgi-fcgi)

**Database:** mysql 10.1.26

**PHP version:** 5.6.33
Modules loaded: Core, date, ereg, libxml, pcre, Reflection, SPL, hash, session, cgi-fcgi, standard, recode, apcu, bcmath, bz2, calendar, ctype, curl, dba, dbase, dom, mbstring, fileinfo, filter, ftp, gd, geoip, gettext, gmp, gnupg, mysqlnd, iconv, imagick, imap, intl, jsmin, json, ldap, exif, mcrypt, mssql, mysql, mysqli, ncurses, odbc, openssl, pcntl, PDO, pgsql, posix, pspell, readline, SimpleXML, snmp, soap, sockets, sqlite3, ssh2, sysvmsg, sysvsem, sysvshm, tidy, tokenizer, xml, xmlwriter, yaml, zip, zlib, mailparse, mongodb, PDFlib, pdo_dblib, pdo_mysql, pdo_pgsql, pdo_sqlite, Phar, scrypt, syck, wddx, xmlreader, xmlrpc, xsl, mhash, apc, Zend OPcache

**Nextcloud version:** 13.0.0 - 13.0..10

**Updated from an older Nextcloud/ownCloud or fresh install:** 

**Where did you install Nextcloud from:** updater

<details><summary>Signing status</summary>

Array
</details>

<details><summary>List of activated apps</summary>

```
Enabled:
 - activity: 2.6.1
 - admin_audit: 1.3.0
 - calendar: 1.6.0
 - comments: 1.3.0
 - contacts: 2.0.1
 - dav: 1.4.6
 - federatedfilesharing: 1.3.1
 - federation: 1.3.0
 - files: 1.8.0
 - files_external: 1.4.1
 - files_pdfviewer: 1.2.0
 - files_sharing: 1.5.0
 - files_texteditor: 2.5.1
 - files_trashbin: 1.3.0
 - files_versions: 1.6.0
 - files_videoplayer: 1.2.0
 - firstrunwizard: 2.2.1
 - gallery: 18.0.0
 - issuetemplate: 0.3.0
 - keeweb: 0.4.0
 - logreader: 2.0.0
 - lookup_server_connector: 1.1.0
 - nextcloud_announcements: 1.2.0
 - notifications: 2.1.2
 - oauth2: 1.1.0
 - password_policy: 1.3.0
 - provisioning_api: 1.3.0
 - serverinfo: 1.3.0
 - sharebymail: 1.3.0
 - survey_client: 1.1.0
 - systemtags: 1.3.0
 - theming: 1.4.1
 - twofactor_backupcodes: 1.2.3
 - updatenotification: 1.3.0
 - workflowengine: 1.3.0
Disabled:
 - bruteforcesettings
 - encryption
 - files_accesscontrol
 - socialsharing_email
 - user_external
 - user_ldap

```
</details>

<details><summary>Configuration (config/config.php)</summary>

```
{
    "datadirectory": "***REMOVED SENSITIVE VALUE***",
    "dbtype": "mysql",
    "version": "13.0..10",
    "installedat": "1334086349.5631",
    "lastupdatedat": "1334087083.9888",
    "dbname": "***REMOVED SENSITIVE VALUE***",
    "dbhost": "***REMOVED SENSITIVE VALUE***",
    "dbtableprefix": "oc_",
    "dbuser": "***REMOVED SENSITIVE VALUE***",
    "dbpassword": "***REMOVED SENSITIVE VALUE***",
    "installed": true,
    "instanceid": "***REMOVED SENSITIVE VALUE***",
    "maxZipInputSize": 419430400,
    "allowZipDownload": true,
    "theme": "",
    "maintenance": false,
    "loglevel": 3,
    "trusted_domains": [
        "owncloud.***.ch",
        "www.***.ch",
        "***.ch"
    ],
    "mail_from_address": "***REMOVED SENSITIVE VALUE***",
    "mail_smtpmode": "php",
    "mail_domain": "***REMOVED SENSITIVE VALUE***",
    "secret": "***REMOVED SENSITIVE VALUE***",
    "trashbin_retention_obligation": "auto, auto",
    "appstore.experimental.enabled": true,
    "memcache.local": "\\OC\\Memcache\\APCu",
    "updater.release.channel": "stable",
    "overwrite.cli.url": "https:\/\/owncloud.***.ch",
    "updater.secret": "***REMOVED SENSITIVE VALUE***"
}
```
</details>

### PR DESCRIPTION
**Pull request by joshtrichards**
Don't do any write operations on files or direcories within RDI.

This will make the Updater more robust in FreeBSD and NFS environments (possibly others but those are the known ones).

Fixes #158

Avoids the problem that arose in #517 due to #515 by keeping the memory usage minimal by doing all the exclusions first, not directly converting the iterator to an array, and so the the resulting array sizes stay minimal.

Verified to work on FreeBSD+NFS web hosting provided by https://hostpoint.sh 

Verified to work with 72,000+ objects in the `apps/` folder (600+ directories containing 120 files each) without exceeding a `memory_limit` of 256M.

What this does:

- basically what we *already* do in `recursiveDelete()`
- most critical in `deleteOldFiles()` & `moveWithExclusions()`

Notes:

- probably would be wise to do in `createBackup()` & `checkWritePermissions()` too, but less critical since deletions/renames aren't taking place in those
- There are opportunities to do refactoring/streamlining/code clean-up but out of scope to keep this PR focused and easily reviewable.

Refs:

- https://bugs.php.net/bug.php?id=69291 (particularly the last comment)
- https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=57696

Cc: @caco3

### CODE REVIEW

**Comment by joshtrichards** [comment]
Test failure in master is unrelated. It's due to nextcloud/server#49224 and currently impacting all PRs in this repo.

**Comment by come-nc** [comment]
Superseeded by https://github.com/nextcloud/updater/pull/516 I think?

**Comment by caco3** [comment]
> Superseeded by #516 I think?

Yes, I would say so.
